### PR TITLE
 New Post : Hashing and Hash Tables

### PR DESCRIPTION
**Pull request by niranjansy**
This article gives an introduction to how hash tables are implemented in programming languages.

### CODE REVIEW

File: blog_src/_posts/2019-10-03-hashing-and-hash-tables.md [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+```cpp
+//L - Linked list, whose each node is the pair (time, IP)
+UpdateAccessList(Log, i, L):  
+//All IPs that accessed service at this instant are appended at the end of the list
+    while Log[i].time <= Now():
+//Append() function adds a node at the end of the list
+        L.Append(Log[i])
+        i = i + 1
+//IPs that had accessed service an hour back are popped from the front of the list
```
Comment (salman-bhai) [inline blog_src/_posts/2019-10-03-hashing-and-hash-tables.md]: Normally convention for writing comments involves having a space after the `//`

File: blog_src/_posts/2019-10-03-hashing-and-hash-tables.md [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+![Formula](/blog/assets/img/Hashing-and-hash-tables/IP_Formula.JPG)
+
+We can create an array A of size 2<sup>32</sup>, where *A[i]* stores the number of times the service was accessed from the IP address, which corresponds to the integer *i*, in the last hour. The following pseudocode implements this idea:
+
+```cpp
+/*Log is the array of log lines to be processed, having pairs (time, IP)
+C is array of counters, of size 2^32
+i - index of first unprocessed log line
+j - index of oldest line in the current 1 hour window*/
```
Comment (salman-bhai) [inline blog_src/_posts/2019-10-03-hashing-and-hash-tables.md]: Umm can you rewrite this as:
```cpp
/*
Log is the array of log lines to be processed, having pairs (time, IP)
C is array of counters, of size 2^32
i - index of first unprocessed log line
j - index of oldest line in the current 1 hour window
*/
```

File: blog_src/_posts/2019-10-03-hashing-and-hash-tables.md [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        j = j - 1
+```
+
+Using this data structure *C*, we can answer all of the above queries conveniently in constant time. But this method has a disadvantage. Even if only a few IPs accessed service, we need to store memory for all the 2<sup>32</sup> possible IP addresses. This wastes a lot of memory. Also we might get instances where the number of objects to be mapped is much more than the memory capacity of the computer, for example mapping all possible strings in the English alphabet. In such cases, this method is not at all feasible, and we need a different method to store the mappings.
+
+#### List Based Mapping
+
+In the previous method of direct addressing, we saw that we stored all possible IPs and this requires too much memory. So one way of solving this is to store only the active IPs, i.e. only the IPs that have accessed service in the last hour. We can implement this using a linked list, sorted in the order of the time of access. Every time an IP becomes active, we append it to the end of the linkd list, and the IPs that are at the beginning of the list, which were accessed more than an hour ago, are removed.
+
+```cpp
```
Comment (salman-bhai) [inline blog_src/_posts/2019-10-03-hashing-and-hash-tables.md]: Not sure if this is cpp code 
If it is pseudo code no need to add cpp to it

File: blog_src/_posts/2019-10-03-hashing-and-hash-tables.md [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+#### Direct Addressing
+
+An IP address consists of four 8-bit integers (i.e. in the range 0-255) separated by dots. Thus there can be a total of 2<sup>32</sup> different IP addresses and so we can convert each IP into a 32-bit integer with a one-to-one mapping. This can be done using the following formula: 
+
+![Formula](/blog/assets/img/Hashing-and-hash-tables/IP_Formula.JPG)
+
+We can create an array A of size 2<sup>32</sup>, where *A[i]* stores the number of times the service was accessed from the IP address, which corresponds to the integer *i*, in the last hour. The following pseudocode implements this idea:
+
+```cpp
```
Comment (salman-bhai) [inline blog_src/_posts/2019-10-03-hashing-and-hash-tables.md]: Don't think this is CPP code

File: blog_src/_posts/2019-10-03-hashing-and-hash-tables.md [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+![Chaining](/blog/assets/img/Hashing-and-hash-tables/Chaining2.JPG)
+
+We will now see the pseudocode for implementing a hash table using chaining.
+
+```cpp
+/*Assume a hash function h : S --> {0, 1, 2, ... m-1}
+Let O be an object in S and V be a mapped value
+A is the array of m lists(chains) of pairs (O,V)
+*/
```
Comment (salman-bhai) [inline blog_src/_posts/2019-10-03-hashing-and-hash-tables.md]: Rewrite these comments as mentioned above!

File: blog_src/_posts/2019-10-03-hashing-and-hash-tables.md [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+*index* = *val* % *m*<br>
+*index* = (*val* + 1) % *m*<br>
+*index* = (*val* + 2) % *m*<br>
+*index* = (*val* + 3) % *m*<br>
+
+Of course, this method works only if the number of keys to be stored in the hash table is less than or equal to the size of the hash table.
+
+**Implementation of Linear Probing :**
+
+```cpp
```
Comment (salman-bhai) [inline blog_src/_posts/2019-10-03-hashing-and-hash-tables.md]: Same as above!

**Comment by niranjansy** [comment]
I have made the required changes to the blog.
(The last commit was made from @adityachirania 's laptop.)

**Pull request closed without merging**